magic-gds-drc/run_all_drc.py

```diff
@@ -0,0 +1,191 @@
+        error_name = error[0]
+        if error_name in acceptable_errors_list:
+            continue
+        errors.append((error[0], error[1:]))
+    
+    return errors
+
+def drc_gds(path: str) -> Tuple[str, List[DRCError]]:
+    """
+    Takes a GDS path. Returns the name of the cell and returns a list of
+    DRC errors.
+    """
+    cell_name = os.path.basename(path)[:-4]
+    
+    env = os.environ.copy()
+    env["PDKPATH"] = PDK_SUBSET
+
+    res = subprocess.run([
+        "python3",
+        STANDARD_DRC_SCRIPT,
+        path
+    ], env=env, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    
+    report_path = path[:-4] + "_drc.txt"
+    try:
+        report = open(report_path).read()
+
+        return cell_name, parse_drc_report(report)
+    except FileNotFoundError:
+        return cell_name, [
+            (
+                "Magic did not produce a report.",
+                [res.stdout.decode("utf8"), res.stderr.decode("utf8")]
+            )
+        ]
+
+@click.command()
+@click.option(
+    "-a",
+    "--acceptable-errors",
+    default="/dev/null",
+    help="A file containing a list of newline-delimited acceptable DRC errors."
+        " Default: No file will be read and all errors deemed unacceptable."
+)
+@click.option(
+    "-m",
+    "--match-directories",
+    default=".",
+    help="A regex that will match subdirectories under cells/."
+        " Default: . (matches everything.)"
+)
+@click.option(
+    "-b",
+    "--known-bad",
+    default="",
+    help="A comma,delimited list of cells that are known bad and"
+        " thus do not cause a non-zero exit upon failure."
+        " Default: empty string (None of them.)"
+)
+def run_all_drc(acceptable_errors, match_directories, known_bad):
+    print("Testing cells in directories matching /%s/…" % match_directories)
+    
+    acceptable_errors_str = open(acceptable_errors).read()
+    global acceptable_errors_list
+    acceptable_errors_list = acceptable_errors_str.split("\n")
+
+    known_bad_list = known_bad.split(",")
+
+    nproc = os.cpu_count()
+    with fut.ThreadPoolExecutor(max_workers=nproc) as executor:
+        futures = []
+
+        cells_dir = "./cells"
+        cells = os.listdir(cells_dir)
+
+        for cell in cells:
+            if not re.match(match_directories, cell):
+                print("Skipping directory %s…" % cell)
+                continue
+
+            cell_dir = os.path.join(cells_dir, cell)
+
+            gds_list = list(
+                filter(lambda x: x.endswith(".gds"), os.listdir(cell_dir))
+            )
+
+            for gds_name in gds_list:
+                gds_path = os.path.join(cell_dir, gds_name)
+
+                futures.append(executor.submit(drc_gds, gds_path))
+
+        successes = 0
+        total = 0
+        exit_code = 0
+        for future in futures:
+            total += 1
+            cell_name, errors = future.result()
+            symbol = "❎"
```

Comment:
Use a red X :-)

Reply:
But then it won't match the tick! :(

Reply:
https://emojipedia.org/hollow-red-circle/ btw